### ISSUE TITLE
core: String Type Categories

### ISSUE THREAD

**Issue by seeyabye** [Jackett/Jackett#9746]
### TLDR
The current Jackett Category structure builds upon the idea of using the category's ID as a unique identifier. This method doesn't work so well when the categories aren't unique to integers, but rather strings.

### Problem Overview
Current categories are added as mappings via 
```C#
AddCategoryMapping(int trackerCategory, TorznabCategory newznabCategory, string trackerCategoryDesc = null)
```

In some cases, trackers use string mappings instead (AnimeBytes and GazelleGamers)

```C#
AddCategoryMapping(string trackerCategory, TorznabCategory newznabCategory, string trackerCategoryDesc = null)
```

The main problem though is behind the logic of `AddCategoryMapping` which tries to parse the string into an integer. It then tries to check if the mapping has already been added to the list of categories. In some cases, this works fine, but when trackers rely on query strings to be strings (i.e. `category[]=Music&category[]=Apps`), we have no way to formulate this correctly.

Currently, trackers that require such query strings are **not** presenting their categories correctly. For instance, on GazelleGamers, the mappings for PC games are added as follows:

```C#
AddCategoryMapping("DOS", TorznabCatType.PCGames, "DOS");
AddCategoryMapping("Windows", TorznabCatType.PCGames, "Windows");
```

Here, you will noticed that `TorznabCatType.PCGames` is repeated twice. With the current logic of `AddCategoryMapping`, only 'DOS' category will be added because both 'DOS' and 'Windows' share the same unique category ID (Specifically, `TorznabCatType.PCGame`).

In other words, there's no real way to add additional categories without a unique category ID. Modifying `TorznabCatType` is not ideal because it should follow the specification set by torznab/newsznab. We shouldn't be adding anything that's not in the specification or it will cause huge inconsistencies. Currently, I think Jackett's category listing follows closely to nZEDb's specification, [https://github.com/nZEDb/nZEDb/blob/master/docs/newznab_api_specification.txt](https://github.com/nZEDb/nZEDb/blob/master/docs/newznab_api_specification.txt).

### The solution

There are a couple of ways to go about this with each solution having its own caveat. These are a few solutions that I could cook up.

1. Have each indexer define their own custom category mapping that subclasses `TorznabCatType`.
This would be the most straightforward solution, but it's left to each indexer to adhere to this principle.

2. Automatically generate unique ID's for each new string categories being added via `AddCategoryMapping`.
There is a consequence to this since some indexers are using custom IDs to map to category IDs on the tracker. Automatically generating IDs for string categories makes sense only when ID is not important. The uniquely generated ID should be adhere to the newsznab specification of having ID > 100000.

At this point of writing and with only a handful of C# indexers having this problem, I am thinking of going with option 1.

This may be related to https://github.com/Jackett/Jackett/issues/8049 in some way or another.

What do you think? @ngosang @garfield69

**Comment by garfield69** [comment]
I don't know much about the innards of category mappings in c# indexers, but what I do know is that we have many yaml indexers that use strings as categories and that those indexers support category search filtering and results category mappings adequately, as far as I know.
So perhaps a place to look is at the yaml processor https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/CardigannIndexer.cs and how it supports such indexers as https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Definitions/zooqle.yml
It may be that the support has always been there but that the particular c# indexer you are researching is not using it.
Then again I may be completely off base ;-)

**Comment by seeyabye** [comment]
Thanks @garfield69. I took a look, at the cardigannindexer, and I think it's more clearer if point it out in yaml indexers as well.

```yaml
  categorymappings:
    # Anime
    - {id: 28, cat: TV/Anime, desc: "Anime/Anime"}
    - {id: 78, cat: TV/Anime, desc: "Anime/Dual Audio"}
    - {id: 79, cat: TV/Anime, desc: "Anime/Dubbed"}
    - {id: 80, cat: TV/Anime, desc: "Anime/Subbed"}
    - {id: 81, cat: TV/Anime, desc: "Anime/Raw"}
```

If we take a look at https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Definitions/1337x.yml, you will see that the category mappings are specified as `TV/Anime` five times. In this example, it's perfectly fine because `AddCategoryMappings` will convert those `id` to custom categories. This is possible because each id is an integer.

Now, what if integers are not supported by the indexer?

```yaml
  categorymappings:
    # Anime
    - {id: "cat[]=anime", cat: TV/Anime, desc: "Anime/Anime"}
    - {id: "cat[]=anime_dual", cat: TV/Anime, desc: "Anime/Dual Audio"}
    - {id: "cat[]=anime_dubbed", cat: TV/Anime, desc: "Anime/Dubbed"}
    - {id: "cat[]=anime_subbed", cat: TV/Anime, desc: "Anime/Subbed"}
    - {id: "cat[]=anime_raw", cat: TV/Anime, desc: "Anime/Raw"}
```
In this case, `AddCategoryMappings` will only add the first id, `cat[]=anime` and the others will be discarded.

As for https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Definitions/zooqle.yml,

If we were to add any additional category like so,

```yaml
caps:
  categories:
    "anime": TV/Anime
    "anime_dubbed": TV/Anime
    "app": PC
    "book": Books
    "files": Other/Misc
    "game": PC/Games
    "movies": Movies
    "music": Audio
    "other": Other
    "tv": TV
```

`anime_dubbed` will not show up, even though it's unique. 

I haven't quite gone through the yaml indexers to see if such use cases exist, but it definitely exist for C# indexers.

**Comment by garfield69** [comment]
Hmm. I think I begin to understand what you are saying.
But I think that saying that only the **first** is **added** seems slightly incorrect.

I tested zooqle by redefining the categories as follows
```yaml
  categorymappings:
    - {id: "anime", cat: TV/Anime, desc: "Anime/RAW"}
    - {id: "anime_dubbed", cat: TV/Anime, desc: "Anime/Dubbed"}
    - {id: "app", cat: PC, desc: "Software"}
    - {id: "book", cat: Books, desc: "Ebook"}
    - {id: "files", cat: Other/Misc, desc: "Files"}
    - {id: "game", cat: PC/Games, desc: "Games"}
    - {id: "movies", cat: Movies, desc: "Films"}
    - {id: "music", cat: Audio, desc: "Music"}
    - {id: "other", cat: Other, desc: "Misc"}
    - {id: "tv", cat: TV, desc: "Series"}
```
and checked out the jackett dashboard manual search
and while as you reported, each category type is listed as a filter once
![image](https://user-images.githubusercontent.com/6933240/95052928-356a4d00-074c-11eb-92c4-77230696ca04.png)
doing a search with a category filter set to 5070(TV/Anime) actually generated a search query that included both 5070 cats in the indexer
https://zooqle.com/search?s=dt&v=t&sd=d&q=%20*%20category:anime,anime_dubbed,
so they are both stored in the internal caps table, its just that the dashboard is showing only the first entry in each cat?

so what we need is some variation of the addCategoryMapping that assigns a custom torznab id
```yaml
  categorymappings:
    - {id: "anime", key: 1, cat: TV/Anime, desc: "Anime/RAW"}
    - {id: "anime_dubbed", key: 2, cat: TV/Anime, desc: "Anime/Dubbed"}
    - {id: "app", key: 3, cat: PC, desc: "Software"}
    - {id: "book", key: 4, cat: Books, desc: "Ebook"}
    - {id: "files", key: 5, cat: Other/Misc, desc: "Files"}
    - {id: "game", key: 6, cat: PC/Games, desc: "Games"}
    - {id: "movies", key: 7, cat: Movies, desc: "Films"}
    - {id: "music", key: 8, cat: Audio, desc: "Music"}
    - {id: "other", key: 9, cat: Other, desc: "Misc"}
    - {id: "tv", cat: key: 10, TV, desc: "Series"}
```
such that internally these will slot into the torznab custom ids 100001 - 100010 to be shown in the dashboard filter list, but are referenced via their string ids when it comes time to build a query.

Feel free to disregard the above if any of it does not make sense and is not useful.

**Comment by seeyabye** [comment]
Indeed you're right. I wasn't aware that the sending query was still retained. In this case, the fix can be a little simpler. I will take a look into this.

**Comment by ngosang** [comment]
I never reviewed the Categories code in detail but there are issues for sure. In this class there are several methods to parse categories with TODOs, duplicate code, try/catch without handling exceptions... => https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/BaseIndexer.cs#L598

1. Before changing the code we must understand how it's working now. I think writing some unit tests for the methods in `BaseIndexer.cs` class are a good way to gain understanding.
2. Move the category logic into a separate class.
3. Refactor the current code without breaking the tests ;)
4. Add new code to solve this issue and #8049 for all trackers.

About the 2 solutions in the first post, well, I don't like them, but I will reconsider after I understand the code. 1) the solution should be generic for all trackers ( I think its doable) 2) we don't need to generate unique ids, the site is providing unique ids in string format. why we need integers instead of strings? I know Torznab handles integers but that is the output, not the internal category handled by us.

I don't have much time during the week, maybe the weekend. You can start without me, I will be here for code review and small questions.

**Comment by ngosang** [comment]
I added some tests for categories here => #9821 There are several issues.

**Comment by garfield69** [comment]
Jackett 0.16.2010
